firefox_puppeteer/ui/windows.py

```diff
@@ -0,0 +1,486 @@
+
+        :returns: The `window handle` of the focused chrome window.
+        """
+        with self.marionette.using_context('chrome'):
+            return self.marionette.execute_script("""
+              Cu.import("resource://gre/modules/Services.jsm");
+              var win = Services.wm.getMostRecentWindow("");
+              return win.QueryInterface(Ci.nsIInterfaceRequestor)
+                        .getInterface(Ci.nsIDOMWindowUtils)
+                        .outerWindowID.toString();
+            """)
+
+    def close(self, handle):
+        """Closes the chrome window with the given handle.
+
+        :param handle: The handle of the chrome window
+        """
+        self.switch_to(handle)
+
+        # TODO: Maybe needs to wait as handled via an observer
+        return self.marionette.close_chrome_window()
+
+    def close_all(self, exceptions=None):
+        """Closes all open chrome windows.
+
+        There is an optional `exceptions` list, which can be used to exclude
+        specific chrome windows from being closed.
+
+        :param exceptions: Optional, list or a single entry of handles or
+         :class:`BaseWindow` instances not to close
+        """
+        handles_to_keep = exceptions or []
+        if not isinstance(handles_to_keep, list):
+            handles_to_keep = [handles_to_keep]
+
+        # Ensure we only have handles and no BaseWindow entries
+        handles_to_keep = [entry.handle if isinstance(entry, BaseWindow) else entry
+                           for entry in handles_to_keep]
+
+        # Find handles to close and close them all
+        handles_to_close = set(self.marionette.chrome_window_handles) - set(handles_to_keep)
+        for handle in handles_to_close:
+            self.close(handle)
+
+    def create_window_instance(self, handle, expected_class=None):
+        """Creates a :class:`BaseWindow` instance for the given chrome window
+
+        :param handle: The handle of the chrome window
+        :param expected_class: Optional, check for the correct window class
+        """
+        current_handle = self.marionette.current_chrome_window_handle
+        window = None
+
+        try:
+            # Retrieve window type to determine the type of chrome window
+            if handle != self.marionette.current_chrome_window_handle:
+                self.switch_to(handle)
+            window_type = self.marionette.get_window_type()
+        finally:
+            # Ensure to switch back to the original window
+            if handle != current_handle:
+                self.marionette.switch_to_window(current_handle)
+
+        if window_type == 'navigator:browser':
+            window = BrowserWindow(lambda: self.marionette, handle)
+        else:
+            raise errors.UnknownWindowError('Unknown window type "%s" for handle: "%s"' %
+                                            (window_type, handle))
+
+        if expected_class is not None and type(window) is not expected_class:
```

Comment:
Something about this doesn't work out to me. We only ever allow creation of BrowserWindow s in this method, so if we ever have expected_class set to something other than BrowserWindow it would fail. Can't individual tests just assert what kind of window type they expect?